src/include/messenger/messenger.h

```diff
@@ -0,0 +1,183 @@
+ *
+ * @see messenger.cpp for a crash course on ZeroMQ, the current backing implementation.
+ */
+class Messenger : public common::DedicatedThreadTask {
+ public:
+  /** Builtin callbacks useful for testing. */
+  enum class BuiltinCallback : uint8_t { NOOP = 0, ECHO, NUM_BUILTIN_CALLBACKS };
+
+  /**
+   * All messages can take in a callback function to be invoked when a reply is received.
+   * Arg 1  :   std::string_view, sender identity.
+   * Arg 2  :   std::string_view, the message itself.
+   */
+  using CallbackFn = std::function<void(std::string_view, std::string_view)>;
+
+  /** @return The default TCP endpoint for a Messenger on the given port. */
+  static ConnectionDestination GetEndpointTCP(std::string target_name, uint16_t port);
+  /** @return The default IPC endpoint for a Messenger on the given port. */
+  static ConnectionDestination GetEndpointIPC(std::string target_name, uint16_t port);
+  /** @return The default INPROC endpoint for a Messenger on the given port. */
+  static ConnectionDestination GetEndpointINPROC(std::string target_name, uint16_t port);
+
+  /**
+   * Create a new Messenger, listening to the default endpoints on the given port.
+   * @param port        The port that determines the default endpoints.
+   * @param identity    The identity that this Messenger instance is known by. See warning!
+   *
+   * @warning           Identity must be unique across all instances of Messengers.
+   */
+  explicit Messenger(uint16_t port, std::string identity);
+
+  /** An explicit destructor is necessary because of the unique_ptr around a forward-declared type. */
+  ~Messenger() override;
+
+  /** Run the main server loop, which dispatches messages received to the MessengerLogic layer. */
+  void RunTask() override;
+
+  /** Terminate the Messenger. */
+  void Terminate() override;
+
+  /**
+   * Listen for new connections on the specified target destination.
+   *
+   * @warning           TODO(WAN): figure out what bad things happen if you give it a ConnectionDestination that is
+   *                     already in use. I don't think this is a problem that is likely to occur because all our
+   *                     destinations are known at compile time and we don't have too many right now, but I should
+   *                     fix this at some point. I am reluctant to add a set of destinations just for this though..
+   *
+   * @param target      The destination to listen on for new connections.
+   */
+  void ListenForConnection(const ConnectionDestination &target);
+
+  /**
+   * Connect to the specified target destination.
+   *
+   * @param target      The destination to be connected to.
+   * @return            A new ConnectionId. See warning!
+   *
+   * @warning           DO NOT USE THIS ConnectionId FROM A DIFFERENT THREAD THAN THE CALLER OF THIS FUNCTION!
+   *                    Make a new connection instead, connections are cheap.
+   */
+  ConnectionId MakeConnection(const ConnectionDestination &target);
```

Comment:
I think it would be more helpful if we have some examples of these? e.g. Correct ordering to use MakeConnection and ListenForConnection; if two messengers are communicating with each other, should both be calling `ListenForConnection` and `MakeConnection`?

Reply:
You shouldn't need `ListenForConnection` at all. Referring to noisepage design slides, that is if you want it to be available on a new endpoint. You can just `MakeConnection` to the default endpoints as shown in this example.

Do you have concrete examples of what you want to see as examples?

Reply:
Oh, I see, I was a little bit confused by the function name. I thought it was listening for connections from some remote destination.